Expose autograd.graph.Node as an abstract base class

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #90105
* #92348
* #91667
* __->__ #91475
* #85849


This PR:
- registers all of the codegened Nodes to the torch._C._functions module, this is where special nodes like AccumulateGrad are already registered.
- creates a autograd.graph.Node abstract base class that all of the newly registered nodes subclass from. We make the subclassing happen by implementing the ``__subclasshook__`` method
- enables static type checking to work and also enables Sphinx to generate documentation for the Node and its methods
- handles both the custom Function and codegened cases

soulitzer committed:
Expose autograd.graph.Node as an abstract base class

[ghstack-poisoned]

soulitzer committed:
Update on "Expose autograd.graph.Node as an abstract base class"




[ghstack-poisoned]

soulitzer commented on docs/source/autograd.rst:
```diff
@@ -259,8 +259,24 @@ Anomaly detection
+Autograd exposes methods to allow one to inspect the graph and interpose behavior during
+the backward pass.
+
+Calling the ``.grad_fn()`` method on a Tensor that is the output of a operation returns an
```
not a method

soulitzer commented on torch/_C/__init__.pyi.in:
```diff
@@ -3,6 +3,7 @@
 import torch
 from torch.package import PackageExporter
 from torch import Tensor
+from torch.autograd.graph import Node as _Node
```
`Node` is already taken by a JIT use case in the scope

soulitzer committed:
Update on "Expose autograd.graph.Node as an abstract base class"


TODO:
- expand on documentation (mainly for hooks registration)
- add tests

[ghstack-poisoned]

soulitzer committed:
Update on "Expose autograd.graph.Node as an abstract base class"


TODO:
- expand on documentation (mainly for hooks registration)
- add tests

[ghstack-poisoned]

soulitzer committed:
Update on "Expose autograd.graph.Node as an abstract base class"



This PR:
- registers all of the codegened Nodes to the torch._C._functions module, this is where special nodes like AccumulateGrad are already registered.
- creates a autograd.graph.Node abstract base class that all of the newly registered nodes subclass from. We make the subclassing happen by implementing the ``__subclasshook__`` method
- enables static type checking to work and also enables Sphinx to generate documentation for the Node and its methods

Haven't thought about what we should do for autograd.Function yet if we decide to do anything.

[ghstack-poisoned]

albanD commented on test/test_autograd.py:
```diff
@@ -6039,6 +6039,21 @@ def backward(ctx, g):
+    def test_autograd_node_isinstance(self):
+        a = torch.rand(3, 3, requires_grad=True)
+        b = a.exp()
+        self.assertTrue(isinstance(b.grad_fn, torch.autograd.graph.Node))
```
You can use assertIsInstance and assertNotIsInstance

albanD commented on test/test_autograd.py:
```diff
@@ -6039,6 +6039,21 @@ def backward(ctx, g):
+        self.assertFalse(isinstance(None, torch.autograd.graph.Node))
+        self.assertFalse(isinstance(1, torch.autograd.graph.Node))
+
+        # isinstance(abc.ABC, abc.ABC) is also False
```
Why?
Looking at `collections.abc`, the classes there return true for this!

soulitzer replied:
I think you mean `issubclass`. Yes, will add checks for those!

albanD commented on test/test_autograd.py:
```diff
@@ -6039,6 +6039,21 @@ def backward(ctx, g):
+        b = a.exp()
+        self.assertTrue(isinstance(b.grad_fn, torch.autograd.graph.Node))
+
+        # This is expected, because it's a type
```
You can use `issubclass()` to check at the type level.

albanD commented on torch/autograd/graph.py:
```diff
@@ -5,15 +5,121 @@
+
+    @property
+    @abc.abstractmethod
+    def next_functions(self) -> Tuple[Tuple[Any, int], ...]:
```
The first element in the inner tuple is an optional Node no?

soulitzer replied:
Yes, but since this method is inside the definition of the class, the class has not been defined yet, so I'm unable to refer to it there. (Maybe there is a better way?)

albanD replied:
You can use string to do that in python. For example in nn.Module we have `def add_module(self, name: str, module: Optional['Module']) -> None:`

soulitzer replied:
very cool, TIL!

albanD commented on torch/autograd/graph.py:
```diff
@@ -5,15 +5,121 @@
+        ...
+
+    @classmethod
+    def __subclasshook__(cls, C):
```
Another option here is to do the same as the colletions.abc classes: they have a register() method that should be used to add new subclasses.
We can then codegen to register all the Nodes and also custom Functions.

But if you think this check is more reliable, I'm happy to keep it.
Does it covers the Nodes defined in `csrc/autograd/functions/` like DelayedError?

soulitzer replied:
I think `register` and `__subclasshook__` do the same thing. The difference is that `register` must be done at initialization time. I don't need to codegen anything because I've already registered all the relevant classes to the `torch._C._function` module, so one solution is to iterate through the classes there.

I prefer the current solution because it is lazier. Instead of registering everything in `torch._C._function` on initialization. Only the people who do the isinstance check have to pay for the hasattr look up on the `torch._C._function` module.

> Does it covers the Nodes defined in csrc/autograd/functions/ like DelayedError?

Yes test has been added now. 

> We can then codegen to register all the Nodes and also custom Functions.

To support custom Function, I'm checking for BackwardCFunction

albanD reviewed (commented):
Very nice!

soulitzer committed:
Update on "Expose autograd.graph.Node as an abstract base class"



This PR:
- registers all of the codegened Nodes to the torch._C._functions module, this is where special nodes like AccumulateGrad are already registered.
- creates a autograd.graph.Node abstract base class that all of the newly registered nodes subclass from. We make the subclassing happen by implementing the ``__subclasshook__`` method
- enables static type checking to work and also enables Sphinx to generate documentation for the Node and its methods

Haven't thought about what we should do for autograd.Function yet if we decide to do anything.

[ghstack-poisoned]

albanD reviewed (approved):
Small things but SGTM otherwise.

soulitzer committed:
Update on "Expose autograd.graph.Node as an abstract base class"



This PR:
- registers all of the codegened Nodes to the torch._C._functions module, this is where special nodes like AccumulateGrad are already registered.
- creates a autograd.graph.Node abstract base class that all of the newly registered nodes subclass from. We make the subclassing happen by implementing the ``__subclasshook__`` method
- enables static type checking to work and also enables Sphinx to generate documentation for the Node and its methods

Haven't thought about what we should do for autograd.Function yet if we decide to do anything.

[ghstack-poisoned]

soulitzer committed:
Update on "Expose autograd.graph.Node as an abstract base class"



This PR:
- registers all of the codegened Nodes to the torch._C._functions module, this is where special nodes like AccumulateGrad are already registered.
- creates a autograd.graph.Node abstract base class that all of the newly registered nodes subclass from. We make the subclassing happen by implementing the ``__subclasshook__`` method
- enables static type checking to work and also enables Sphinx to generate documentation for the Node and its methods

Haven't thought about what we should do for autograd.Function yet if we decide to do anything.

[ghstack-poisoned]

soulitzer committed:
Update on "Expose autograd.graph.Node as an abstract base class"



This PR:
- registers all of the codegened Nodes to the torch._C._functions module, this is where special nodes like AccumulateGrad are already registered.
- creates a autograd.graph.Node abstract base class that all of the newly registered nodes subclass from. We make the subclassing happen by implementing the ``__subclasshook__`` method
- enables static type checking to work and also enables Sphinx to generate documentation for the Node and its methods
- handles both the custom Function and codegened cases


[ghstack-poisoned]

soulitzer committed:
Update on "Expose autograd.graph.Node as an abstract base class"



This PR:
- registers all of the codegened Nodes to the torch._C._functions module, this is where special nodes like AccumulateGrad are already registered.
- creates a autograd.graph.Node abstract base class that all of the newly registered nodes subclass from. We make the subclassing happen by implementing the ``__subclasshook__`` method
- enables static type checking to work and also enables Sphinx to generate documentation for the Node and its methods
- handles both the custom Function and codegened cases


[ghstack-poisoned]

soulitzer committed:
Update on "Expose autograd.graph.Node as an abstract base class"



This PR:
- registers all of the codegened Nodes to the torch._C._functions module, this is where special nodes like AccumulateGrad are already registered.
- creates a autograd.graph.Node abstract base class that all of the newly registered nodes subclass from. We make the subclassing happen by implementing the ``__subclasshook__`` method
- enables static type checking to work and also enables Sphinx to generate documentation for the Node and its methods
- handles both the custom Function and codegened cases


[ghstack-poisoned]
